Introduce multi-release JAR

jasontedor opened the pull request:
This commit introduces the ability for the core Elasticsearch JAR to be a multi-release JAR containing code that is compiled for JDK 8 and code that is compiled for JDK 9. At runtime, a JDK 8 JVM will ignore the JDK 9 compiled classfiles, and a JDK 9 JVM will use the JDK 9 compiled classfiles instead of the JDK 8 compiled classfiles. With this work, we utilize the new JDK 9 API for obtaining the PID of the running JVM, instead of relying on a hack.

For now, we want to keep IDEs on JDK 8 so when the build is in an IDE we ignore the JDK 9 source set (as otherwise the IDE would give compilation errors). However, with this change, running Gradle from the command-line now requires JAVA_HOME and JAVA_9_HOME to be set. This will require follow-up work in our CI infrastructure and our release builds to accommodate this change.

jasontedor committed:
Introduce multi-release JAR

This commit introduces the ability for the core Elasticsearch JAR to be
a multi-release JAR containing code that is compiled for JDK 8 and code
that is compiled for JDK 9. At runtime, a JDK 8 JVM will ignore the JDK
9 compiled classfiles, and a JDK 9 JVM will use the JDK 9 compiled
classfiles instead of the JDK 8 compiled classfiles. With this work, we
utilize the new JDK 9 API for obtaining the PID of the running JVM,
instead of relying on a hack.

For now, we want to keep IDEs on JDK 8 so when the build is in an IDE we
ignore the JDK 9 source set (as otherwise the IDE would give compilation
errors). However, with this change, running Gradle from the command-line
now requires JAVA_HOME and JAVA_9_HOME to be set. This will require
follow-up work in our CI infrastructure and our release builds to
accommodate this change.

nik9000 commented on core/src/main/java/org/elasticsearch/monitor/jvm/JvmPid.java:
```diff
@@ -0,0 +1,50 @@
+
+import java.lang.management.ManagementFactory;
+
+class JvmPid {
```
Many of our test fixtures could use this too.

nik9000 commented on core/src/main/java9/org/elasticsearch/monitor/jvm/JvmPid.java:
```diff
@@ -0,0 +1,32 @@
+class JvmPid {
+
+    static long getPid() {
+        return ProcessHandle.current().pid();
```
Nice!

nik9000 commented on core/build.gradle:
```diff
@@ -36,6 +36,30 @@ publishing {
+  sourceSets {
+    java9 {
+      java {
+        srcDirs = ['src/main/java9']
```
Neat.

nik9000 reviewed (approved):
I like it. Probably good to have @rjernst look too.

I admit to not having tried it in Eclipse mostly because I haven't used Eclipse for a few months now.

jasontedor committed:
Remove debugging imports

jasontedor committed:
Merge branch 'master' into mrjar

* master:
  Enable convert processor to support Long and Double. (#27957)
  Enable Wildfly tests on JDK 9 and JDK 10
  [Test] Fix allowed delta for calculated scores in DiscountedCumulativeGainTests

jasontedor committed:
Require JDK 9 for compilation

This commit modifies the build to require JDK 9 for
compilation. Henceforth, we will compile with a JDK 9 compiler targeting
JDK 8 as the class file format. Optionally, JAVA_8_HOME can be set as
the runtime JDK used for running tests. To enable this change, we
separate the meaning of the compiler Java home versus the runtime Java
home. If the runtime Java home is not set (via JAVA_8_HOME) then we
fallback to using JAVA_HOME as the runtime Java home. This enables:
 - developers only have to set one Java home (JAVA_HOME)
 - developers can set an optional Java home (JAVA_8_HOME) to test on the
   minimum supported runtime
 - we can test compiling with JDK 9 running on JDK 8 and compiling with
   JDK 9 running on JDK 9 in CI

jasontedor committed:
Update docs

jasontedor committed:
Fix reindex from remote tests

jasontedor committed:
Merge branch 'master' into compile-with-jdk-9

* master:
  Set the elasticsearch-nio codebase for tests (#28067)
  Bump compat version for local depdendent test to 6.2.0
  Pass `java.locale.providers=COMPAT` to Java 9 onwards (#28080)
  Allow shrinking of indices from a previous major (#28076)
  Remove deprecated exceptions (#28059)
  Add Writeable.Reader support to TransportResponseHandler (#28010)
  Plugins: Add plugin extension capabilities (#27881)

jasontedor committed:
Merge branch 'master' into compile-with-jdk-9

* master:
  test: replaced try-catch statements with expectThrows(...)
  Add getWarmer and getTranslog method to NodeIndicesStats (#28092)
  fix doc mistake
  Added ASN support for Ingest GeoIP plugin.
  Fix global aggregation that requires breadth first and scores (#27942)
  Introduce Gradle wrapper
  Ignore GIT_COMMIT when calculating commit hash
  Re-enable bwc tests after #27881 was backported

jasontedor committed:
Runtime Java home

jasontedor committed:
Merge branch 'master' into mrjar

* master: (25 commits)
  Remove Gradle cheatsheet
  Fix reproduction info to point to Gradle wrapper
  Update platforms tests to use Gradle wrapper
  Update testing docs to reflect Gradle wrapper
  Painless: Modify Loader to Load Classes Directly from Definition (#28088)
  Update contributing docs to use the Gradle wrapper
  Create nio-transport plugin for NioTransport (#27949)
  test: replaced try-catch statements with expectThrows(...)
  Add getWarmer and getTranslog method to NodeIndicesStats (#28092)
  fix doc mistake
  Added ASN support for Ingest GeoIP plugin.
  Fix global aggregation that requires breadth first and scores (#27942)
  Introduce Gradle wrapper
  Ignore GIT_COMMIT when calculating commit hash
  Re-enable bwc tests after #27881 was backported
  Set the elasticsearch-nio codebase for tests (#28067)
  Bump compat version for local depdendent test to 6.2.0
  Pass `java.locale.providers=COMPAT` to Java 9 onwards (#28080)
  Allow shrinking of indices from a previous major (#28076)
  Remove deprecated exceptions (#28059)
  ...

jasontedor committed:
Merge branch 'compile-with-jdk-9' into mrjar

* compile-with-jdk-9:
  Runtime Java home
  Fix reindex from remote tests
  Update docs
  Require JDK 9 for compilation

jasontedor committed:
Merge branch 'master' into compile-with-jdk-9

* master:
  Remove Gradle cheatsheet
  Fix reproduction info to point to Gradle wrapper
  Update platforms tests to use Gradle wrapper
  Update testing docs to reflect Gradle wrapper
  Painless: Modify Loader to Load Classes Directly from Definition (#28088)
  Update contributing docs to use the Gradle wrapper
  Create nio-transport plugin for NioTransport (#27949)

jasontedor committed:
Merge branch 'master' into compile-with-jdk-9

* master:
  Use Gradle wrapper when building BWC
  Painless: Add a simple cache for whitelist methods and fields. (#28142)
  Fix upgrading indices which use a custom similarity plugin. (#26985)
  Fix Licenses values for CDDL and Custom URL (#27999)
  Cleanup TcpChannelFactory and remove classes (#28102)
  Fix expected plugins test for transport-nio
  [Docs] Fix Date Math example descriptions (#28125)
  Fail rollover if duplicated alias found in template (#28110)
  Avoid concurrent snapshot finalizations when deleting an INIT snapshot (#28078)
  Deprecate `isShardsAcked()` in favour of `isShardsAcknowledged()` (#27819)
  [TEST] Wait for replicas to be allocated before shrinking
  Use the underlying connection version for CCS connections  (#28093)
  test: do not use asn fields
  Test: Add assumeFalse for test that cannot pass on windows
  Clarify reproduce info on Windows
  Remove out-of-date projectile file

jasontedor committed:
Use JDK 8 if available when building legacy

jasontedor committed:
Merge branch 'master' into compile-with-jdk-9

* master:
  Fix Gradle wrapper usage on Windows when building BWC (#28146)
  [Docs] Fix some typos in comments (#28098)

jasontedor committed:
Merge branch 'master' into compile-with-jdk-9

* master:
  Set watermarks in single-node test cases
  Add the ability to bundle multiple plugins into a meta plugin (#28022)
  Declare empty package dirs as output dirs
  Consistent updates of IndexShardSnapshotStatus (#28130)

jasontedor committed:
Merge branch 'master' into compile-with-jdk-9

* master: (30 commits)
  Fix lock accounting in releasable lock
  Add ability to associate an ID with tasks  (#27764)
  [DOCS] Removed differencies between text and code (#27993)
  text fixes (#28136)
  Update getting-started.asciidoc (#28145)
  [Docs] Spelling fix in painless-getting-started.asciidoc (#28187)
  Fixed the cat.health REST test to accept 4ms, not just 4.0ms (#28186)
  Do not keep 5.x commits once having 6.x commits (#28188)
  Rename core module to server (#28180)
  upgraded jna from 4.4.0-1 to 4.5.1 (#28183)
  [TEST] Do not call RandomizedTest.scaledRandomIntBetween from multiple threads
  Primary send safe commit in file-based recovery (#28038)
  [Docs] Correct response json in rank-eval.asciidoc
  Add scroll parameter to _reindex API (#28041)
  Include all sentences smaller than fragment_size in the unified highlighter (#28132)
  Modifies the JavaAPI docs related to AggregationBuilder
  [Docs] Improvements in script-fields.asciidoc (#28174)
  [Docs] Remove Kerberos/SPNEGO Shield plugin (#28019)
  Ignore null value for range field (#27845) (#28116)
  Fix environment variable substitutions in list setting (#28106)
  ...

jasontedor committed:
Merge branch 'master' into compile-with-jdk-9

* master:
  TEST: init unassigned gcp in testAcquireIndexCommit
  Replica start peer recovery with safe commit (#28181)
  Truncate tlog cli should assign global checkpoint (#28192)

jasontedor committed:
Merge branch 'compile-with-jdk-9' into mrjar

* compile-with-jdk-9: (56 commits)
  TEST: init unassigned gcp in testAcquireIndexCommit
  Replica start peer recovery with safe commit (#28181)
  Truncate tlog cli should assign global checkpoint (#28192)
  Fix lock accounting in releasable lock
  Add ability to associate an ID with tasks  (#27764)
  [DOCS] Removed differencies between text and code (#27993)
  text fixes (#28136)
  Update getting-started.asciidoc (#28145)
  [Docs] Spelling fix in painless-getting-started.asciidoc (#28187)
  Fixed the cat.health REST test to accept 4ms, not just 4.0ms (#28186)
  Do not keep 5.x commits once having 6.x commits (#28188)
  Rename core module to server (#28180)
  upgraded jna from 4.4.0-1 to 4.5.1 (#28183)
  [TEST] Do not call RandomizedTest.scaledRandomIntBetween from multiple threads
  Primary send safe commit in file-based recovery (#28038)
  [Docs] Correct response json in rank-eval.asciidoc
  Add scroll parameter to _reindex API (#28041)
  Include all sentences smaller than fragment_size in the unified highlighter (#28132)
  Modifies the JavaAPI docs related to AggregationBuilder
  [Docs] Improvements in script-fields.asciidoc (#28174)
  ...

jasontedor committed:
Merge branch 'master' into mrjar

* master: (35 commits)
  Move the multi-get response tests to server
  Require JDK 9 for compilation (#28071)
  Revert "[Docs] Fix Java Api index administration usage (#28133)"
  Revert "[Docs] Fix base directory to include for put_mapping.asciidoc"
  Added multi get api to the high level rest client.
  [Docs] Clarify numeric datatype ranges (#28240)
  [Docs] Fix base directory to include for put_mapping.asciidoc
  Open engine should keep only starting commit (#28228)
  [Docs] Fix Java Api index administration usage (#28133)
  Fix eclipse build. (#28236)
  Never return null from Strings.tokenizeToStringArray (#28224)
  Fallback to TransportMasterNodeAction for cluster health retries (#28195)
  [Docs] Changes to ingest.asciidoc (#28212)
  TEST: Update logging for testAckedIndexing
  [GEO] Add WKT Support to GeoBoundingBoxQueryBuilder
  Painless: Add whitelist extensions (#28161)
  Fix daitch_mokotoff phonetic filter to use the dedicated Lucene filter (#28225)
  Avoid doing redundant work when checking for self references. (#26927)
  Fix casts in HotThreads. (#27578)
  Ignore the `-snapshot` suffix when comparing the Lucene version in the build and the docs. (#27927)
  ...
